src/backend/distributed/planner/modify_planner.c

```diff
@@ -641,3 +644,315 @@ ExtractPartitionValue(Query *query, Var *partitionColumn)
 
 	return partitionValue;
 }
+
+
+/*
+ * ContainsPartitionColumnMatchExpression returns true if the query contains an exact
+ * match (equal) expression on the partition column. The function returns true only
+ * if the match expression has an AND relation with the rest of the expression tree.
+ */
+static bool
+ContainsPartitionColumnMatchExpression(Node *node, Var *partitionColumn)
+{
+	if (node == NULL)
+	{
+		return  false;
+	}
+
+	if (IsA(node, OpExpr))
+	{
+		OpExpr *opExpr = (OpExpr *) node;
+		bool simpleExpression = SimpleOpExpression((Expr *)opExpr);
+		bool partitionColumnInExpr = false;
+		char *operatorName  = NULL;
+		int equalsCompare = -1;
+		bool usingEqualityOperator = false;
+
+		if (!simpleExpression)
+		{
+			return false;
+		}
+
+		partitionColumnInExpr = OpExpressionContainsColumn(opExpr, partitionColumn);
+
+		if (!partitionColumnInExpr)
+		{
+			return false;
+		}
+
+		operatorName = get_opname(opExpr->opno);
+		equalsCompare = strncmp(operatorName, EQUAL_OPERATOR_STRING, NAMEDATALEN);
+		usingEqualityOperator = (equalsCompare == 0);
+		if (usingEqualityOperator)
+		{
+			return true;
+		}
+	}
+	else if (IsA(node, BoolExpr))
+	{
+		BoolExpr *boolExpr = (BoolExpr *) node;
+
+		if (boolExpr->boolop == AND_EXPR)
+		{
+			List *argumentList = boolExpr->args;
+			ListCell *argumentCell = NULL;
+
+			foreach(argumentCell, argumentList)
+			{
+				Node *argumentNode = (Node *) lfirst(argumentCell);
+				bool partitionColumnMatch =
+					ContainsPartitionColumnMatchExpression(argumentNode, partitionColumn);
+
+				if (partitionColumnMatch)
+				{
+					return true;
+				}
+			}
+		}
+	}
+
+	return false;
+}
+
+
+/*
+ * MultiRouterPlannableQuery returns true if given query can be router plannable.
+ * The query is router plannable if it is a select query issued on a hash
+ * partitioned distributed table, and it has a exact match comparison on the
+ * partition column. This feature is enabled if task executor is set to real-time or
+ * router.
+*/
+bool
+MultiRouterPlannableQuery(Query *query)
+{
```

Comment:
Having and approximate distinct does not add much value here. However an inconsistency is possible. One runs a router plannable query with approximate distinct on. Gets the exact distinct. Then he runs the same query, still going to the same shard, but due to some change in where clauses query becomes not-router plannable, goes through other path. returns approximate distincts. Query result is different than the previous one.